main/keyword.c

```diff
@@ -256,3 +256,14 @@ extern void dumpKeywordTable (FILE *fp)
 		}
 	}
 }
+
+extern void addKeywordGroup (struct keywordGroupDefinition *groupdef,
+							 langType language, bool checkDuplication)
+{
+	for (int i = 0; groupdef->keywords[i]; i++)
+	{
+		if ((!checkDuplication)
+			|| (lookupKeyword (groupdef->keywords[i], language) == KEYWORD_NONE))
```

Comment:
Even if `checkDuplication` is false, adding a keyword twice is not a good idea; maybe it is a bug.
I can put Assert here in the case that `checkDuplication` is false.

Reply:
> Even if checkDuplication is false, adding a keyword twice is not a good idea; maybe it is a bug.

When I used this routine, I expected that new value overwrote the old value.  I think it is more natural behavior. Now checkDuplication gives us freedom to chose the behavior.

I think we should not add Assert.

Reply:
addKeywordGroup() has two usages.
One is found in (System)Verilog parser: adding keywords that are not added yet.
One will be found in Vala: grouping keywords by their values.

In the former case, trying to add a keyword more than once is o.k; Assert is not needed.
In the latter case, trying to add a keyword implies a programming bug; Assert is needed.

I updated my change, as I wrote above.

Reply:
> One is found in (System)Verilog parser: adding keywords that are not added yet.
> One will be found in Vala: grouping keywords by their values.
> 
> In the former case, trying to add a keyword more than once is o.k; Assert is not needed.
> In the latter case, trying to add a keyword implies a programming bug; Assert is needed.

I agree with you here.

In the former case: `checkDuplication `is set to false. Assert is not needed.
In the latter case: `checkDuplication `is set to true. Assert is needed.

> I can put Assert here in the case that checkDuplication is false.

You wrote opposite.

I've checked-out your commit.  But I don't see `checkDuplication`. Did you stop to add `checkDuplication`?

Reply:
I found I've been misunderstanding the issue.

What I expected is adding `checkDuplication` argument to `addKeyword()` to stop Assert().

`addKeyword()` is now use only by c.c and verilog.c.  Adding an argument can be acceptable.
How do you think?

Reply:
> What I expected is adding checkDuplication argument to addKeyword() to stop Assert().

I think I should move the Assert in addKeywordGroup() to addKeyword(). I would like to keep addKeyword() as is.

> addKeyword() is now use only by c.c and verilog.c. Adding an argument can be acceptable.
How do you think?

addKeyword is used in main/parse.c, too. 

I have changed the code of ctags more declarative style since u-ctags started.
Ideally, a parser should just declare keywords. Calling addKeyword is imperative.
If possible, I would like to avoid calling it. However, c.c and verilog.c must call it because more than one parser are defined in a source file. They build their keyword tables from shared data structures with addKeyword().
I added addKeywordGroup() for making the way of building more declarative.

Reply:
> If possible, I would like to avoid calling it. 

This makes sense to me.  I assumed it was a public API.